### PR TITLE
8302073: Specifying OnError handler prevents WatcherThread to break a deadlock in report_and_die()

### PR DESCRIPTION
**Pull request by apavlyutkin**
The patch fixes error reporting to check timeout in the case when a user specifies OnError hander. Before VMError:check_timeout() ignored timeout in this case, and so didn't break malloc() deadlock.

Verification (amd64/20.04LTS): the idea of the test is to crash JVM running with error hander of 3 successive `sleep` commands for 1s, 10s, and 60s with and without specified timeout

```
16:52:17@alex@alex-VirtualBox>( echo "
public class C {
  public static void main(String[] args) throws Throwable {
>     while (true) Thread.sleep(1000);
>   }
> }
> " >> C.java )
16:57:35@alex@alex-VirtualBox>./build/linux-x86_64-server-release/images/jdk/bin/java -XX:OnError='sleep 1;sleep 10;sleep 60' ./C.java &
[2] 179574
17:00:19@alex@alex-VirtualBox>kill -s SIGSEGV 179574
17:00:27@alex@alex-VirtualBox>#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f7b1701ecd5 (sent by kill), pid=179574, tid=179574
#
# JRE version: OpenJDK Runtime Environment (21.0) (build 21-internal-adhoc.alex.jdk)
# Java VM: OpenJDK 64-Bit Server VM (21-internal-adhoc.alex.jdk, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# C  [libpthread.so.0+0x9cd5]  __pthread_clockjoin_ex+0x255
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport %p %s %c %d %P %E" (or dumping to /home/alex/jdk/core.179574)
#
# An error report file with more information is saved as:
# /home/alex/jdk/hs_err_pid179574.log
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#
#
# -XX:OnError="sleep 1;sleep 10;sleep 60"
#   Executing /bin/sh -c "sleep 1" ...
#   Executing /bin/sh -c "sleep 10" ...
#   Executing /bin/sh -c "sleep 60" ...

[2]+  Aborted                 (core dumped) ./build/linux-x86_64-server-release/images/jdk/bin/java -XX:OnError='sleep 1;sleep 10;sleep 60' ./C.java
17:02:03@alex@alex-VirtualBox>./build/linux-x86_64-server-release/images/jdk/bin/java -XX:ErrorLogTimeout=5 -XX:OnError='sleep 1;sleep 10;sleep 60' ./C.java &
[2] 179602
17:02:32@alex@alex-VirtualBox>kill -s SIGSEGV 179602
17:02:41@alex@alex-VirtualBox>#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f9d71b18cd5 (sent by kill), pid=179602, tid=179602
#
# JRE version: OpenJDK Runtime Environment (21.0) (build 21-internal-adhoc.alex.jdk)
# Java VM: OpenJDK 64-Bit Server VM (21-internal-adhoc.alex.jdk, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# C  [libpthread.so.0+0x9cd5]  __pthread_clockjoin_ex+0x255
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport %p %s %c %d %P %E" (or dumping to /home/alex/jdk/core.179602)
#
# An error report file with more information is saved as:
# /home/alex/jdk/hs_err_pid179602.log
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#
#
# -XX:OnError="sleep 1;sleep 10;sleep 60"
#   Executing /bin/sh -c "sleep 1" ...
#   Executing /bin/sh -c "sleep 10" ...

------ Timeout during error reporting after 11 s. ------

17:02:54@alex@alex-VirtualBox>
```

Regression (amd64/20.04LTS): `test/hotspot/jtreg/runtime/ErrorHandling` with different combinations of `-vmoption:-XX:ErrorLogTimeout=10` and `-vmoption:-XX:OnError='sleep 10'`

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302073](https://bugs.openjdk.org/browse/JDK-8302073): Specifying OnError handler prevents WatcherThread to break a deadlock in report_and_die() ⚠️ Issue is not open.

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12925/head:pull/12925` \
`$ git checkout pull/12925`

Update a local copy of the PR: \
`$ git checkout pull/12925` \
`$ git pull https://git.openjdk.org/jdk.git pull/12925/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12925`

View PR using the GUI difftool: \
`$ git pr show -t 12925`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12925.diff">https://git.openjdk.org/jdk/pull/12925.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12925#issuecomment-1460219300)

### CODE REVIEW

**Review by tstuefe — COMMENTED** [review]
Thinking this through some more, I'm starting to doubt we do the right thing here. It is certainly convoluted: we *are* the reporting thread here, so what happens is that:
- after fork_and_exec, we call check_timeout
- check_timeout will signal ourself
- we receive the signal
- we enter the secondary signal handler recursively
- we then re-enter VMError::report_and_die
- we then note that this is a timeout and print an error message and call os::die.

That is too complicated for my taste, and I don't know if there are any hidden issues with VM::check_timeout(). Since now, we call it from two threads, possibly, the reporting thread and the watcher thread. That function was not intended for concurrent usage.

And before thinking about the correct behavior, we need to clarify if the protection we grant an OnError invocation extends to the whole chain of error scripts. Right now we say OnError scripts should not be interrupted. Okay, but what about the next OnError script? If the user specifies several OnError scripts, should they all get a chance to run to finish? 

Because denying the follow-up error scripts a chance to run feels weirdly arbitrary. Either those scripts are essential, or they aren't. If they are, all should run. If they are not, it should be okay to kill the JVM *while it is waiting for the child process to finish* - this would make the patch simpler, and people argue that this would be the correct behavior anyway.

Personally, I think that killing the JVM while it is in waitpid waiting for the child is probably benign. Child would be reparented, possibly zombified on badly set up systems, but that's it. It would probably run to completion.

**Comment by tstuefe** [comment]
Pinging @dholmes-ora for input and opinion

File: src/hotspot/share/utilities/vmError.cpp [hunk 1]
```diff
@@ -1356,6 +1356,24 @@ void VMError::report_and_die(Thread* thread, const char* filename, int lineno, s
...
 }
 
+namespace {
+  class ForkAndExecCheckPoint : public StackObj {
+    NONCOPYABLE( ForkAndExecCheckPoint );
+    static int _in_progress;
+  public:
+    ForkAndExecCheckPoint() {
+      assert(Atomic::load(&_in_progress) == 0, "fork_and_exec() is already in progress");
+      Atomic::store(&_in_progress, 1);
```
[the remark this answers is not in the document]
Comment (apavlyutkin) [reply src/hotspot/share/utilities/vmError.cpp:1366]: Initially I used exactly CAS, but CAS causes full memory fence. On other hand report_and_die() seems the last place to care about performance

File: src/hotspot/share/utilities/vmError.cpp [hunk 2]
```diff
@@ -1356,6 +1356,24 @@ void VMError::report_and_die(Thread* thread, const char* filename, int lineno, s
   report_and_die(vm_err_type, nullptr, detail_fmt, detail_args, thread, nullptr, nullptr, nullptr, filename, lineno, size);
 }
 
+namespace {
+  class ForkAndExecCheckPoint : public StackObj {
+    NONCOPYABLE( ForkAndExecCheckPoint );
+    static int _in_progress;
```
[the remark this answers is not in the document]
Comment (apavlyutkin) [reply src/hotspot/share/utilities/vmError.cpp:1362]: sure

File: src/hotspot/share/utilities/vmError.cpp [hunk 3]
```diff
@@ -1356,6 +1356,24 @@ void VMError::report_and_die(Thread* thread, const char* filename, int lineno, s
   report_and_die(vm_err_type, nullptr, detail_fmt, detail_args, thread, nullptr, nullptr, nullptr, filename, lineno, size);
 }
 
+namespace {
+  class ForkAndExecCheckPoint : public StackObj {
```
[the remark this answers is not in the document]
Comment (apavlyutkin) [reply src/hotspot/share/utilities/vmError.cpp:1360]: ok

**Comment by apavlyutkin** [comment]
> Thinking this through some more, I'm starting to doubt we do the right thing here. It is certainly convoluted: we _are_ the reporting thread here, so what happens is that:
> 
> * after fork_and_exec, we call check_timeout
> * check_timeout will signal ourself
> * we receive the signal
> * we enter the secondary signal handler recursively
> * we then re-enter VMError::report_and_die
> * we then note that this is a timeout and print an error message and call os::die.
> 
> That is too complicated for my taste, and I don't know if there are any hidden issues with VM::check_timeout(). Since now, we call it from two threads, possibly, the reporting thread and the watcher thread. That function was not intended for concurrent usage.
> 
> And before thinking about the correct behavior, we need to clarify if the protection we grant an OnError invocation extends to the whole chain of error scripts. Right now we say OnError scripts should not be interrupted. Okay, but what about the next OnError script? If the user specifies several OnError scripts, should they all get a chance to run to finish?
> 
> Because denying the follow-up error scripts a chance to run feels weirdly arbitrary. Either those scripts are essential, or they aren't. If they are, all should run. If they are not, it should be okay to kill the JVM _while it is waiting for the child process to finish_ - this would make the patch simpler, and people argue that this would be the correct behavior anyway.
> 
> Personally, I think that killing the JVM while it is in waitpid waiting for the child is probably benign. Child would be reparented, possibly zombified on badly set up systems, but that's it. It would probably run to completion.

ok, let's replace check_timeout() with something like
```
jlong expiration = get_reporting_start_time() + ( jlong )ErrorLogTimeout * TIMESTAMP_TO_SECONDS_FACTOR;
if ( get_current_timestamp() > expiration ) break;
```

**Comment by jsolomon8080** [comment]
Hi - I'm the originator of this bug report. I'm glad this is getting fixed and I will ultimately defer to the java experts since I have never contributed code to this code base, but if it were me, I'd do the simplest thing possible. The solution presented here seems overly complex. I don't understand why an OnError script or an abort_hook is so special. I think it's up to the user to ensure that neither takes longer than ErrorLogTimeout, which defaults to 2 minutes. That's an eternity. Do java users really expect to wait 2 minutes for their processes to exit?

If this were my code, I would remove any guarantee about OnError and make it responsibility of the user to set ErrorLogTimeout appropriately.

I understand that there may be many users who have counted on this guarantee and you can't break them now. I also understand that I'm not aware of all the ways that OnError is used. I'm sure we will get to some solution that will fix the real problem, which is what I care about the most. Thank you.

**Comment by dholmes-ora** [comment]
I thought we were still discussing options in JBS so this PR seems premature to me. I agree with @tstuefe initial comment this seems way too complex and I'm not even sure I can figure out the control flow here.

**Comment by tstuefe** [comment]
@dholmes-ora My proposal would be to be pragmatic and continue the discussion here. This has been going on too long. Switching channels again would be more confusing. 

I would prefer JBS or ML discussion as well, but not everyone (eg. @jsolomon8080) has write access to JBS.

----

I thought this over some more.

A big part of this problem is a plain bug in VMError::check_timeout(): it should never block *step timeouts*. We added step timeouts back in 2016 to deal with these kinds of problems: if the JVM hangs in error reporting, give it a kick to get it going again. Do so repeatedly. This is related to the global timeout, but works also independently from it (and therefore from the question of when to honor the global timeout).

I opened https://bugs.openjdk.org/browse/JDK-8303861 to deal with this, see PR: https://github.com/openjdk/jdk/pull/12936

This may already be a big help for cases like these. In fact, it may already be enough, and we could maybe close this issue.

However, if we have a recursive malloc situation, we may hang repeatedly. JVM will kick itself alive every time (with my patch) but this is still annoying. The root problem here is that we should not use malloc during error handling. Cannot always be avoided, but at least we should minimize malloc use. 

Decoder, in particular, should not use malloc. Therefore I also opened https://bugs.openjdk.org/browse/JDK-8303862 to track that. I won't have time to work on that. I have the hope that maybe @chhagedorn can :-) ? Otherwise Azul may also chip in some bug fixing.

----

All these are unrelated to the question of whether OnError should be blocked or not. I realize now that if we decide to (continue to) protect OnError from timeouts, we must never act on the global timeout until all OnError steps ran. Since these run right before VM exit, the original implementation that just blocked the global timeout altogether was actually right.

So, there's that. Maybe JDK-8303861 is already enough for cases like this. At least JDK-8303861 does not introduce any backward compatibility issues.

**Comment by tstuefe** [comment]
> Hi - I'm the originator of this bug report. I'm glad this is getting fixed and I will ultimately defer to the java experts since I have never contributed code to this code base, but if it were me, I'd do the simplest thing possible. The solution presented here seems overly complex. I don't understand why an OnError script or an abort_hook is so special. I think it's up to the user to ensure that neither takes longer than ErrorLogTimeout, which defaults to 2 minutes. That's an eternity. Do java users really expect to wait 2 minutes for their processes to exit?

Some do, some don't, there are cases for both. But that is not the point. We must guarantee that all OnError commands are run and that the abort hook is called. Since these things happen after error reporting, we cannot interrupt error reporting.

This is a backward compatibility question. There are arguments for ErrorLogTimeout having precedence, and there are arguments for the current behavior. Both sides have good arguments, but OpenJDK tries to err on the side of compatibility here.

> If this were my code, I would remove any guarantee about OnError and make it responsibility of the user to set ErrorLogTimeout appropriately.

Nothing is stopping Azul from changing this behavior downstream in their VM, but upstream we need to find a consensus with all interested parties.

> 
> I understand that there may be many users who have counted on this guarantee and you can't break them now. I also understand that I'm not aware of all the ways that OnError is used. I'm sure we will get to some solution that will fix the real problem, which I care about the most. Thank you.

Let's see if this issue is moot once the fix for JDK-8303861 is out.

Cheers, Thomas

**Comment by chhagedorn** [comment]
> Decoder, in particular, should not use malloc. Therefore I also opened https://bugs.openjdk.org/browse/JDK-8303862 to track that. I won't have time to work on that. I have the hope that maybe @chhagedorn can :-) ?

Does any of our current allocation strategies allow the usage of a custom scratch buffer (i.e. the error reporting scratch buffer) as memory location? If not, it could get more complicated. I could still try to tackle that but I'm not sure if I have the necessary knowledge in that area.

**Comment by tstuefe** [comment]
> > Decoder, in particular, should not use malloc. Therefore I also opened https://bugs.openjdk.org/browse/JDK-8303862 to track that. I won't have time to work on that. I have the hope that maybe @chhagedorn can :-) ?
> 
> Does any of our current allocation strategies allow the usage of a custom scratch buffer (i.e. the error reporting scratch buffer) as memory location? If not, it could get more complicated. I could still try to tackle that but I'm not sure if I have the necessary knowledge in that area.

What we usually do for these kind of problems is to pass the scratch buffer via argument to the processing functions. And make that use either it or, if NULL was passes, allocate its own thing.

Another way to do this would be to add a scratch buffer pointer to Thread.

A third way to do this (I had been experimenting with it) would be to pre-allocate a scratch buffer, and once error handling began, to use that inside os::malloc. That is the most involved solution, though, and I'm not particularly fond of it. I know many reviewers would hate it, too :-)

**Comment by tstuefe** [comment]
@apavlyutkin @jsolomon8080 https://bugs.openjdk.org/browse/JDK-8303861 is in. Would be interesting to see if this already helps the customer.

**Comment by apavlyutkin** [comment]
> @apavlyutkin @jsolomon8080 https://bugs.openjdk.org/browse/JDK-8303861 is in. Would be interesting to see if this already helps the customer.

Sure, I'll try to backport the changes to zulu11. Thanks a lot

**Comment by chhagedorn** [comment]
> > > Decoder, in particular, should not use malloc. Therefore I also opened https://bugs.openjdk.org/browse/JDK-8303862 to track that. I won't have time to work on that. I have the hope that maybe @chhagedorn can :-) ?
> > 
> > 
> > Does any of our current allocation strategies allow the usage of a custom scratch buffer (i.e. the error reporting scratch buffer) as memory location? If not, it could get more complicated. I could still try to tackle that but I'm not sure if I have the necessary knowledge in that area.
> 
> What we usually do for these kind of problems is to pass the scratch buffer via argument to the processing functions. And make that use either it or, if NULL was passes, allocate its own thing.
> 
> Another way to do this would be to add a scratch buffer pointer to Thread.
> 
> A third way to do this (I had been experimenting with it) would be to pre-allocate a scratch buffer, and once error handling began, to use that inside os::malloc. That is the most involved solution, though, and I'm not particularly fond of it. I know many reviewers would hate it, too :-)

Thanks for the summary. Could we also use "placement new" to use the scratch buffer to allocate and create objects to? For example, when creating a decoder, to use something like
```
decoder = new(scratch_buffer) ElfDecoder();
```
here:
https://github.com/openjdk/jdk/blob/4e631fa43fd821846c12ae2177360c44cf770766/src/hotspot/share/utilities/decoder.cpp#L62-L70
Then we can still utilize polymorphism. But then `AbstractDecoder` needs to be something else than `CHeapObj`. And of course, this approach also needs careful checking that a new object actually fits into the provided scratch buffer. On top of that, it gets more complicated when we want to keep multiple objects alive in the same scratch buffer. So, I'm not sure if this approach is feasible.

**Comment by jsolomon8080** [comment]
> > @apavlyutkin @jsolomon8080 https://bugs.openjdk.org/browse/JDK-8303861 is in. Would be interesting to see if this already helps the customer.
> 
> Sure, I'll try to backport the changes to zulu11. Thanks a lot

I look forward to receiving a version of zulu11 with this change and testing it out!

**Comment by tstuefe** [comment]
> > > > Decoder, in particular, should not use malloc. Therefore I also opened https://bugs.openjdk.org/browse/JDK-8303862 to track that. I won't have time to work on that. I have the hope that maybe @chhagedorn can :-) ?
> > > 
> > > 
> > > Does any of our current allocation strategies allow the usage of a custom scratch buffer (i.e. the error reporting scratch buffer) as memory location? If not, it could get more complicated. I could still try to tackle that but I'm not sure if I have the necessary knowledge in that area.
> > 
> > 
> > What we usually do for these kind of problems is to pass the scratch buffer via argument to the processing functions. And make that use either it or, if NULL was passes, allocate its own thing.
> > Another way to do this would be to add a scratch buffer pointer to Thread.
> > A third way to do this (I had been experimenting with it) would be to pre-allocate a scratch buffer, and once error handling began, to use that inside os::malloc. That is the most involved solution, though, and I'm not particularly fond of it. I know many reviewers would hate it, too :-)
> 
> Thanks for the summary. Could we also use "placement new" to use the scratch buffer to allocate and create objects to? For example, when creating a decoder, to use something like
> 
> ```
> decoder = new(scratch_buffer) ElfDecoder();
> ```
> 
> here:
> 
> https://github.com/openjdk/jdk/blob/4e631fa43fd821846c12ae2177360c44cf770766/src/hotspot/share/utilities/decoder.cpp#L62-L70
> 
> Then we can still utilize polymorphism. But then `AbstractDecoder` needs to be something else than `CHeapObj`. And of course, this approach also needs careful checking that a new object actually fits into the provided scratch buffer. On top of that, it gets more complicated when we want to keep multiple objects alive in the same scratch buffer. So, I'm not sure if this approach is feasible, though.

Its possible to use placement new, and also to place several objects like this; but I realize now the scratch buffer might be too small. 

I need to think this through a bit. Maybe providing a pre-allocated buffer for os::malloc would be the right thing to do.

**Comment by chhagedorn** [comment]
> > > > > Decoder, in particular, should not use malloc. Therefore I also opened https://bugs.openjdk.org/browse/JDK-8303862 to track that. I won't have time to work on that. I have the hope that maybe @chhagedorn can :-) ?
> > > > 
> > > > 
> > > > Does any of our current allocation strategies allow the usage of a custom scratch buffer (i.e. the error reporting scratch buffer) as memory location? If not, it could get more complicated. I could still try to tackle that but I'm not sure if I have the necessary knowledge in that area.
> > > 
> > > 
> > > What we usually do for these kind of problems is to pass the scratch buffer via argument to the processing functions. And make that use either it or, if NULL was passes, allocate its own thing.
> > > Another way to do this would be to add a scratch buffer pointer to Thread.
> > > A third way to do this (I had been experimenting with it) would be to pre-allocate a scratch buffer, and once error handling began, to use that inside os::malloc. That is the most involved solution, though, and I'm not particularly fond of it. I know many reviewers would hate it, too :-)
> > 
> > 
> > Thanks for the summary. Could we also use "placement new" to use the scratch buffer to allocate and create objects to? For example, when creating a decoder, to use something like
> > ```
> > decoder = new(scratch_buffer) ElfDecoder();
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > here:
> > https://github.com/openjdk/jdk/blob/4e631fa43fd821846c12ae2177360c44cf770766/src/hotspot/share/utilities/decoder.cpp#L62-L70
> > 
> > Then we can still utilize polymorphism. But then `AbstractDecoder` needs to be something else than `CHeapObj`. And of course, this approach also needs careful checking that a new object actually fits into the provided scratch buffer. On top of that, it gets more complicated when we want to keep multiple objects alive in the same scratch buffer. So, I'm not sure if this approach is feasible, though.
> 
> Its possible to use placement new, and also to place several objects like this; but I realize now the scratch buffer might be too small.

Yes, that might be the case.

> 
> I need to think this through a bit. Maybe providing a pre-allocated buffer for os::malloc would be the right thing to do.

That might be better and easier to do instead of going with a separate allocation strategy with placement new.

**Pull request closed without merging**